drivers/src/csrng.rs

```diff
@@ -0,0 +1,346 @@
+
+            self.num_words_left -= 1;
+
+            Some(registers().genbits().read())
+        }
+    }
+}
+
+impl ExactSizeIterator for Iter<'_> {
+    fn len(&self) -> usize {
+        self.num_words_left
+    }
+}
+
+impl FusedIterator for Iter<'_> {}
+
+impl Drop for Iter<'_> {
+    fn drop(&mut self) {
+        // Exhaust this generate request.
+        for _ in self {}
+    }
+}
+
+/// Contains counts of failing health checks.
+///
+/// This struct is returned by the [`health_counts`] method on [`Csrng`].
+///
+/// [`health_counts`]: Csrng::health_counts
+pub struct HealthFailCounts {
+    /// The total number of failing health check alerts.
+    pub total: u32,
+
+    /// The counts of specific failing health checks.
+    pub specific: AlertFailCountsReadVal,
+}
+
+fn registers() -> csrng::RegisterBlock {
+    csrng::RegisterBlock::csrng_reg()
+}
+
+fn send_command(command: Command) -> CaliptraResult<()> {
+    // https://opentitan.org/book/hw/ip/csrng/doc/theory_of_operation.html#general-command-format
+    let acmd: u32;
+    let clen: usize;
+    let flag0: MultiBitBool;
+    let glen: usize;
+    let extra_words: &[u32];
+    let err: CsrngErr;
+
+    match command {
+        Command::Instantiate(ref seed) | Command::Reseed(ref seed) => {
+            acmd = if matches!(command, Command::Instantiate(_)) {
+                err = CsrngErr::Instantiate;
+                1
+            } else {
+                err = CsrngErr::Reseed;
+                2
+            };
+
+            match seed {
+                Seed::EntropySrc => {
+                    clen = 0;
+                    flag0 = MultiBitBool::False;
+                    extra_words = &[];
+                }
+
+                Seed::Constant(constant) => {
+                    clen = constant.len().min(MAX_SEED_WORDS);
+                    flag0 = MultiBitBool::True;
+                    extra_words = &constant[..clen];
+                }
+            }
+
+            glen = 0;
+        }
+
+        Command::Generate { num_128_bit_blocks } => {
+            acmd = 3;
+            clen = 0;
+            flag0 = MultiBitBool::False;
+            glen = num_128_bit_blocks.min(MAX_GENERATE_BLOCKS);
+            extra_words = &[];
+            err = CsrngErr::Generate;
+        }
+
+        Command::Update(words) => {
+            acmd = 4;
+            clen = words.len().min(MAX_SEED_WORDS);
+            flag0 = MultiBitBool::True;
+            glen = 0;
+            extra_words = &words[..clen];
+            err = CsrngErr::Update;
+        }
+
+        Command::Uninstantiate => {
+            acmd = 5;
+            clen = 0;
+            flag0 = MultiBitBool::False;
+            glen = 0;
+            extra_words = &[];
+            err = CsrngErr::Uninstantiate;
+        }
+    }
+
+    // Write mandatory 32-bit command header.
+    registers()
+        .cmd_req()
+        .write(|_| ((glen as u32) << 12) | ((flag0 as u32) << 8) | ((clen as u32) << 4) | acmd);
```

Comment:
Can you mask the fields (I know this is probably fine as is with all the constants above, but I'd prefer to err on the safe side)?

Reply:
Done.